Add key window to achieve unique keys for delete operations

joshjms opened the pull request:
Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

Adds the `keyStore` into `etcdTraffic` which maintains a consistent pool of keys (size 10) such that used keys will be swapped out and replaced with a new key. 

```bash
PERSIST_RESULTS=true EXPECT_DEBUG=true GO_TEST_FLAGS='-v' make test-robustness
```

The porcupine results shows that the uniqueness of keys in early stages for EtcdTrafficDeleteLeases:

![image](https://github.com/user-attachments/assets/a44084d4-1a28-4f5c-9229-4df403cbcc7e)

and in the later stages:

![image](https://github.com/user-attachments/assets/092808d4-8ab0-4a69-9c2f-4335cf442ecf)


ref: https://github.com/etcd-io/etcd/issues/19984
/cc @serathius

nwnt commented on tests/robustness/traffic/key_window.go:
```diff
@@ -0,0 +1,62 @@
+	randVal := rand.Intn(100)
+	if randVal < 10 {
+		return false
+	}
```
```suggestion
return rand.Intn(100) >= 10
```

nwnt commented on tests/robustness/traffic/key_window.go:
```diff
@@ -0,0 +1,62 @@
+}
+
+func (k *KeyWindow) useWindow() bool {
+	randVal := rand.Intn(100)
```
This function can be simplified to be something like
```suggestion
return !rand.Intn(100) < 10
```

nwnt commented on tests/robustness/traffic/key_window.go:
```diff
@@ -0,0 +1,62 @@
+func NewKeyWindow(size int, keyPrefix string) *KeyWindow {
+	k := &KeyWindow{
+		Counter:    0,
+		WindowSize: size,
```
can we instead use len(Keys) when we need the size?

serathius replied:
+1

nwnt commented on tests/robustness/traffic/key_window.go:
```diff
@@ -0,0 +1,62 @@
+type KeyWindow struct {
+	sync.Mutex
+
+	Counter    int
```
if we don't use this anywhere, I think let's not make it public.

serathius replied:
All those fields should be private and we should fill them through constructor.

nwnt commented on tests/robustness/traffic/key_window.go:
```diff
@@ -0,0 +1,62 @@
+	}
+
+	// This will use the default key0 to key9
+	randKey := rand.Intn(k.WindowSize)
```
Since `randkey` doesn't seem to be used anywhere, we can combine line 49 and 50 to be:
```go
key := rand.Intn(k.WindowSize)
```

nwnt commented on tests/robustness/traffic/key_window.go:
```diff
@@ -0,0 +1,62 @@
+func (k *KeyWindow) GetKey() string {
+	k.Lock()
+	defer k.Unlock()
+
```
I think we can structure this function by making it like this:
```go
key := rand.Intn(k.WindowSize)
if k.useWindow() {
  randkey := key
  key = k.Keys[randKey]
  k.Keys[randKey] = k.Counter
  k.Counter++
}
return fmt.Sprintf("%s%d", k.keyPrefix, key)
```

nwnt replied:
no hold on one sec on this one.

nwnt replied:
nvm, I think I kinda got it right, confused myself about WindowSize and the random value in useWindow().

serathius commented on tests/robustness/traffic/key_window.go:
```diff
@@ -0,0 +1,62 @@
+	"sync"
+)
+
+type KeyWindow struct {
```
Think we should pick a better name. The `window` in name comes from idea for algorithm where we have a moving window of key range. However, we should rather name it based on it's purpose. Like `keyStore`, `keyGenerator`, `keyProvider`

joshjms replied:
sliding window but nothing is sliding :frowning_face: ok will change to keyStore

serathius commented on tests/robustness/traffic/key_window.go:
```diff
@@ -0,0 +1,62 @@
+
+	Counter    int
+	WindowSize int
+	Keys       []int
```
Maybe Key field should include key names and not numbers used to generated keys.

joshjms commented:
Thanks for the reviews guys, I will address them as soon as possible. Thanks for bearing with me :smile:

joshjms commented:
I addressed all the reviews you guys gave, please let me know if there's anything I missed. Please look at the updated screenshots for the new evidence.

joshjms commented:
/retest

serathius commented on tests/robustness/traffic/etcd.go:
```diff
@@ -215,21 +215,21 @@ type etcdTrafficClient struct {
 }
 
-func (c etcdTrafficClient) Request(ctx context.Context, request etcdRequestType, lastRev int64) (rev int64, err error) {
+func (c etcdTrafficClient) Request(ctx context.Context, request etcdRequestType, lastRev int64, keyStore *keyStore) (rev int64, err error) {
```
You can put the keyStore on `etcdTrafficClient` struct.

serathius commented on tests/robustness/traffic/key_store.go:
```diff
@@ -0,0 +1,105 @@
+	useKeyIndex := rand.Intn(len(k.keys))
+	useKey := k.keys[useKeyIndex]
+
+	if rand.Intn(100) < 90 {
+		k.keys[useKeyIndex] = fmt.Sprintf("%s%d", k.keyPrefix, k.counter)
+		k.counter++
+	}
```
```suggestion
	if rand.Intn(100) < 90 {
		k.keys[useKeyIndex] = fmt.Sprintf("%s%d", k.keyPrefix, k.counter)
		k.counter++
	}
```
Move this to a separate function`replaceKey`

serathius commented on tests/robustness/traffic/key_store.go:
```diff
@@ -0,0 +1,105 @@
+	for i, op := range ops {
+		keys[i] = k.keys[permutedKeyIndexes[i]]
+
+		if op == model.DeleteOperation && rand.Intn(100) < 90 {
```
This repeats rolling 90%

joshjms replied:
I completely missed removing the roll inside, thanks for noticing
